akka-http-core/src/main/scala/akka/http/impl/engine/ws/FrameLogger.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2009-2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.impl.engine.ws
+
+import akka.NotUsed
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.stream.Attributes
+import akka.stream.Attributes.LogLevels
+import akka.stream.scaladsl.{ BidiFlow, Flow }
+import akka.util.ByteString
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[ws] object FrameLogger {
+  final val maxBytes = 16
+
+  def logFramesIfEnabled(shouldLog: Boolean): BidiFlow[FrameEvent, FrameEvent, FrameEvent, FrameEvent, NotUsed] =
+    if (shouldLog) bidi
+    else BidiFlow.identity
+
+  def bidi: BidiFlow[FrameEvent, FrameEvent, FrameEvent, FrameEvent, NotUsed] =
+    BidiFlow.fromFlows(
+      Flow[FrameEvent].log(s"${Console.RED}DOWN${Console.RESET}", FrameLogger.logEvent),
+      Flow[FrameEvent].log(s"${Console.GREEN} UP ${Console.RESET}", FrameLogger.logEvent))
+      .addAttributes(Attributes(LogLevels(Logging.DebugLevel, Logging.DebugLevel, Logging.DebugLevel)))
+
+  def logEvent(frameEvent: FrameEvent): String = {
+    case class LogEntry(frameType: String, data: String)
+
+    def hex(bytes: ByteString): String = {
+      val num = math.min(maxBytes, bytes.size)
+      val ellipsis = if (num < bytes.size) s" [... ${bytes.size - num} more bytes]" else ""
+      bytes
+        .take(num)
+        .map(_ formatted "%02x")
+        .mkString(" ") + ellipsis
+    }
+
+    def entryForFrame(frameEvent: FrameEvent): LogEntry =
+      frameEvent match {
+        case FrameStart(_, data) => LogEntry("FRAMESTART", hex(data))
+        case FrameData(data, _)  => LogEntry("FRAMEDATA", hex(data))
```

Comment:
This event is basically a continuation of the `FrameStart` one (i.e. it has no correspondence with anything in the WS specs, it's more about how frames are represented in akka-http).

The text should indicate that it's a continuation of a previous frame and whether more data is expected. Maybe we could just use ellipsis for that?

Reply:
Are you suggesting to just use the ellipsis for `FrameData` ?

Reply:
Yes, I made that change.